ogame_core/src/resources.rs

```diff
@@ -1,69 +1,182 @@
-use std::ops::{AddAssign, Mul, SubAssign};
+use std::{ops::{AddAssign, Mul, SubAssign, Sub, Add}, collections::{BTreeMap, HashMap}};
 
 use serde::{Deserialize, Serialize};
 
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize, Debug, Hash, Ord, PartialOrd)]
+pub enum ResourceType{
+    Metal,
+    Crystal,
+    Deuterium,
+}
+
+#[derive(Clone, PartialEq, Serialize, Deserialize, Debug)]
+pub struct Resource {
+    pub id: String,
+    pub amount: f64,
+}
+
+impl Resource {
+    pub fn new(amount: f64) -> Self {
+        Resource {
+            id: "".to_string(),
+            amount
+        }
+    }
+
+    pub fn set_id(&mut self, id: String) -> Self {
+        self.id = id;
+        self.clone()
+    }
+
+    pub fn has_enough(&self, other: Self) -> bool {
+        self.amount >= other.amount
+    }
+
+    pub fn pay(&mut self, other: Self) {
+        self.amount -= other.amount;
+    }
+
+    pub fn gain(&mut self, other: Self) {
+        self.amount += other.amount;
+    }
+}
+
 #[derive(Clone, Debug, Serialize, Deserialize, Default)]
 pub struct Resources {
     pub id: String,
-    pub metal: f64,
-    pub crystal: f64,
-    pub deuterium: f64,
+    pub items: BTreeMap<ResourceType, f64>,
+}
+
+impl From<Vec<(ResourceType, f64)>> for Resources {
+    fn from(items: Vec<(ResourceType, f64)>) -> Self {
+        let arr: BTreeMap<ResourceType, f64> = items.into_iter().map(|(resource_type, amount)| {
+            (resource_type, amount)
+        }).collect();
+
+        let items = BTreeMap::from(arr);
```

Comment:
This is pointless if you already have a BTreeMap, no ?
Furthermore, there is an existing From<[(T, U)]> implementation that save a Vec allocation ;)